index.js

```diff
@@ -92,386 +71,170 @@ function printArray(array) {
+    while (!current.done) {
+      var value = print(current.value[1], refs, opts, state);
 
-        result += indentLines(value);
+      result += indent(value, opts);
 
-        current = iterator.next();
+      current = iterator.next();
 
-        if (!current.done) {
-          result += ',\n';
-        }
+      if (!current.done) {
+        result += ',\n';
       }
-
-      result += '\n';
     }
 
-    return result + '}';
-  }
-});
-
-/**
- * @public
- * @class String
- * @extends Type
- * @memberOf Type
- */
-Type.String = new Type({
-  test: _.isString,
-
-  print: function(val) {
-    return '"' + val + '"';
-  }
-});
-
-/**
- * @public
- * @class Symbol
- * @extends Type
- * @memberOf Type
- */
-Type.Symbol = new Type({
-  test: _.isSymbol,
-
-  print: function(val) {
-    return Symbol.prototype.toString.call(val).replace(SYMBOL_REGEXP, 'Symbol($1)');
-  }
-});
-
-/**
- * @public
- * @class Undefined
- * @extends Type
- * @memberOf Type
- */
-Type.Undefined = new Type({
-  test: _.isUndefined,
-
-  print: function() {
-    return 'undefined';
+    result += '\n';
   }
-});
 
-Type.WeakMap = new Type({
-  test: _.isWeakMap,
+  return result + '}';
+}
 
-  print: function() {
-    return 'WeakMap {}';
-  }
-});
+function printValue(val, refs, opts, state) {
+  // Simple values
+  if ( isBoolean   (val) ) return Boolean.prototype.toString.call(val);
+  if ( isDate      (val) ) return Date.prototype.toISOString.call(val);
+  if ( isError     (val) ) return '[' + Error.prototype.toString.call(val) + ']';
+  if ( isFunction  (val) ) return Function.prototype.toString.call(val);
+  if ( isInfinity  (val) ) return Infinity.toString.call(val);
+  if ( isNaN       (val) ) return 'NaN';
+  if ( isNull      (val) ) return 'null';
+  if ( isNumber    (val) ) return isNegativeZero(val) ? '-0' : '' + val;
+  if ( isRegExp    (val) ) return RegExp.prototype.toString.call(val)
+  if ( isString    (val) ) return '"' + val + '"';
+  if ( isSymbol    (val) ) return Symbol.prototype.toString.call(val).replace(SYMBOL_REGEXP, 'Symbol($1)');
+  if ( isUndefined (val) ) return 'undefined';
+  if ( isWeakMap   (val) ) return 'WeakMap {}';
+  if ( isWeakSet   (val) ) return 'WeakSet {}';
+
+  var stop = opts.maxDepth < state.depth;
+
+  if ( isArguments (val) ) return stop ? '[Arguments]' : printArguments (val, refs, opts, state);
+  if ( isArrayish  (val) ) return stop ? '[Array]'     : printArray     (val, refs, opts, state);
+  if ( isMap       (val) ) return stop ? '[Map]'       : printMap       (val, refs, opts, state);
+  if ( isSet       (val) ) return stop ? '[Set]'       : printSet       (val, refs, opts, state);
+  // purposefully last:
+  if ( isObject    (val) ) return stop ? '[Object]'    : printObject    (val, refs, opts, state);
+}
 
-Type.WeakMap = new Type({
-  test: _.isWeakMap,
+function print(val, refs, opts, state) {
+  refs = refs.slice(); // clone
 
-  print: function() {
-    return 'WeakMap {}';
+  if (refs.indexOf(val) !== -1) {
```

Comment:
Is there a way to make a weakset from another weakset? See the line above where I `.slice()` the refs to clone it for the current depth in the tree.